src/linker/Linker/CustomAttributeSource.cs

```diff
@@ -30,13 +28,9 @@ public IEnumerable<CustomAttribute> GetCustomAttributes (ICustomAttributeProvide
 					yield return customAttribute;
 			}
 
-			if (_sources.Count > 0) {
-				foreach (var source in _sources) {
-					if (source.HasCustomAttributes (provider)) {
-						foreach (var customAttribute in source.GetCustomAttributes (provider))
-							yield return customAttribute;
-					}
-				}
+			if (_xmlCustomAttributes.ContainsKey (provider)) {
+				foreach (var customAttribute in _xmlCustomAttributes.TryGetValue (provider, out var ann) ? ann : null)
```

Comment:
It doesn't work with null values, still in this case since it was after a ContainsKey is never going to be null. Still is more performant only running TryGetValue once. So Im going to change it.

Reply:
I end up doing
```
if (_xmlCustomAttributes.ContainsKey (provider)) {
	foreach (var customAttribute in _xmlCustomAttributes[provider])
		yield return customAttribute;
}
```
Which is optimized for not founding the provider, since it never allocates the default in TryGetValue (https://referencesource.microsoft.com/#mscorlib/system/collections/generic/dictionary.cs,2e5bc6d8c0f21e67)
I made a small benchmark and TryGetValue is faster than doing ContainsKey and then trying to use the get indexer when the provider exists. 
So this implementation will be faster if the value is not found but will be slower in the case the value is found. In this case most of the time the value will not be found.

Reply:
Nit: It's good to add a comment into the code with this reasoning - for example I would be tempted to rewrite it with just `TryGet` 😄